### PR TITLE
Use content negotiation in the dynamic client.

### PR DESCRIPTION
**Pull request by benluddy**
#### What type of PR is this?

/kind cleanup
/sig api-machinery

#### What this PR does / why we need it:

Replaces the hardcoded JSON serialization in the dynamic client with the existing RESTClient support for content negotiation. This in itself should not change the observable behavior of the dynamic client but it is a prerequisite to supporting additional encodings (like CBOR).

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by benluddy — COMMENTED** [review]
/cc @deads2k

File: staging/src/k8s.io/client-go/dynamic/scheme.go [hunk 1]
```diff
@@ -106,3 +100,24 @@ func (t unstructuredTyper) ObjectKinds(obj runtime.Object) ([]schema.GroupVersio
 func (t unstructuredTyper) Recognizes(gvk schema.GroupVersionKind) bool {
 	return true
 }
+
+// The dynamic client formerly hardcoded object marshaling and unmarshaling and would allow
+// Unstructured objects that were missing apiVersion and/or kind in requests.
```
Comment (deads2k) [inline staging/src/k8s.io/client-go/dynamic/scheme.go:105]: allow in serializing requests or in deserializing responses?

**Comment by deads2k** [comment]
It's going to be a pain, but I'd like see a test-server created with unit tests that save the content received from the client for the cases to be sure the content and headers are the same.  We do this in bootstrap RBAC by serializing JSON for the unit tests and using an env var to write that content.

Similarly, a unit test with a test server to be sure the golang API returns the same result for various cases of response bytes.

File: staging/src/k8s.io/client-go/dynamic/scheme.go (hunk 1 above)
Comment (benluddy) [reply staging/src/k8s.io/client-go/dynamic/scheme.go:105]: Serializing requests. Directly encoding with `(unstructuredJSONScheme).Encode` is effectively a pass-through to `json.Marshal`. With this PR, request objects now pass through a `WithVersionEncoder` which fails on missing kind errors from https://github.com/kubernetes/kubernetes/blob/b722d017a34b300a2284b890448e5a605f21d01e/staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go#L246-L253.

**Comment by benluddy** [comment]
No problem, I will add those tests.

File: staging/src/k8s.io/client-go/dynamic/scheme.go (hunk 1 above)
Comment (deads2k) [inline staging/src/k8s.io/client-go/dynamic/scheme.go:105]: Based on testing, these were serialized, but then failed when used for a create, update, or apply.  so we're changing from a server-side error after the request to a client-side error just before the request, correct?

Comment (deads2k) [reply staging/src/k8s.io/client-go/dynamic/scheme.go:105]: > Based on testing, these were serialized, but then failed when used for a create, update, or apply. so we're changing from a server-side error after the request to a client-side error just before the request, correct?

Nearly correct.  CRs fail, but built-in types accept json without kind or apiVersion.

**Comment by deads2k** [comment]
/kind cleanup

It looks smaller, not sure it's easier to debug in the end, but I can see how the refactor makes the content negotiation work.  Thank you for adding the test a while back to prove that this change is zero-diff for output.

/lgtm
/approve

**Pull request merged**